Tests/MongoSwiftSyncTests/UnifiedTestRunner/Matching.swift

```diff
@@ -0,0 +1,311 @@
+import MongoSwiftSync
+import TestsCommon
+
+/// Enum representing types that can be matched against expected values.
+enum MatchableResult {
+    /// A root document. i.e. a documents where extra keys are ignored when matching against an expected document.
+    case rootDocument(BSONDocument)
+    /// An array of root documents.
+    case rootDocumentArray([BSONDocument])
+    /// A (non-root) document.
+    case document(BSONDocument)
+    /// An array of BSONs.
+    case array([BSON])
+    /// A non-document, non-array BSON.
+    case scalar(BSON)
+    /// A nil result.
+    case none
+
+    /// Initializes an instance of `MatchableResult` from a `BSON`.
+    init(from bson: BSON?) {
+        guard let bson = bson else {
+            self = .none
+            return
+        }
+        switch bson {
+        case let .document(doc):
+            self = .document(doc)
+        case let .array(arr):
+            self = .array(arr)
+        default:
+            self = .scalar(bson)
+        }
+    }
+}
+
+extension UnifiedOperationResult {
+    /// Determines whether this result matches `expected`.
+    func matches(expected: BSON, entities: EntityMap) throws -> Bool {
+        let actual: MatchableResult
+        switch self {
+        case let .bson(bson):
+            actual = MatchableResult(from: bson)
+        case let .rootDocument(doc):
+            actual = .rootDocument(doc)
+        case let .rootDocumentArray(arr):
+            actual = .rootDocumentArray(arr)
+        case .none:
+            actual = .none
+        default:
+            return false
+        }
+
+        return try matchesInner(
+            expected: expected,
+            actual: actual,
+            entities: entities
+        )
+    }
+}
+
+/// Determines whether `actual` matches `expected`, recursing if needed for nested documents and arrays.
+private func matchesInner(
```

Comment:
Now that we have a `MatchableResult` type, what about adding this as a method to it like we added a method to `UnifiedOperationResult`?